New flow for data deletion and editorial improvements (#629)
italia/eidas-it-wallet-docs: docs/en/wallet-solution.rst

This section lists the requirements that Wallet Providers, Wallet Solutions, and their Wallet Instances must meet.

- The Wallet Solution MUST adhere to the specifications set by this document for obtaining Personal Identification (PID) and (Q)EAAs.
- The Wallet Provider MUST expose a set of endpoints, exclusively available to its Wallet Solution instances, supporting the core functionalities of the Wallet Instances.
- The Wallet Instance MUST periodically reestablish trust with its Wallet Provider, obtaining a fresh Wallet Attestation.
- The Wallet Instance MUST establish trust with other participants of the Wallet ecosystem, such as Credential Issers and Relying Parties, presenting a Wallet Attestation.
- The Wallet Instance MUST be compatible and functional on both Android and iOS operating systems and available on the Play Store and App Store, respectively.
- The Wallet Instance MUST provide a mechanism to verify the User's actual possession and full control of their personal device.

> - The Wallet Instance MUST be able to present the an up-to-date list of relying parties with which the User has established a connection and, where applicable, all data exchanged;

Reviewer: ```suggestion
- The Wallet Instance MUST provide Users with an up-to-date list of Relying Parties with which the User has established a connection and, where applicable, all data exchanged;
```

> - The Wallet Instance MUST be able to allow users to request the erasure of personal attributes by a Relying Party pursuant to Article 17 of Regulation (EU) 2016/679, and to log each Erasure Request made.

Reviewer: ```suggestion
- The Wallet Instance MUST provide Users with a mechanism to request the erasure of personal attributes by a Relying Party pursuant to Article 17 of Regulation (EU) 2016/679, and to log each Erasure Request made.
```

Wallet Instance Initialization and Registration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This process allows the User who has just installed the Wallet Instance application to register the Wallet Instance with the Wallet Provider backend. During this process, the Wallet Instance application will request a security and integrity assertion from the OS manufacturer, which also binds a long-lived key pair stored in a proper secure storage within the device itself. This assertion will be validated by the Wallet Provider, and if the validation is successful, the Wallet Provider will authenticate the Wallet Instance. For details see :ref:`Mobile Application Instance Initialization`.

.. note::

>   The Wallet Provider SHOULD associate the Wallet Instance (through the ``hardware_key_tag`` identifier) with a specific User uniquely identified within the Wallet Provider's systems. This will be useful for the lifecycle of the Wallet Instance and for a future revocation. For details see 

Reviewer: the sentence is incomplete

Author: This change is part of my commit.

.. include:: wallet-attestation.rst
.. include:: wallet-revocation.rst

User's Attributes Deletion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> This Wallet Instance functionality allows the User to obtain from the Instance itself a list of all Relying Parties towards which attributes that can uniquely identify Users (e.g., the Codice Fiscale claim of the PID) have been presented. Subsequently the User is free to request deletion of all attributes presented to a Relying Party of its choice. Below the high level flow regarding this interaction is presented.   

Reviewer: ```suggestion
This Wallet Instance functionality allows Users to obtain a list of all Relying Parties towards which attributes that can uniquely identify Users (e.g., the tax_id_code claim of the PID) have been presented. Subsequently Users may request deletion of all attributes presented to a Relying Party of their choice. Below the high level flow regarding this interaction is presented.   
```

.. figure:: ../../images/user's_data_deletion_flow.svg
    :figwidth: 100%
    :align: center
    :target: https://www.plantuml.com/plantuml/png/RLBDZjD03BxdAQozS67t0Nf0ksABn0KX5iI5YvCuxOZfE8mzBNrxx2bDcBAtTFpv-_7NHr7CMWwnmwASog6dtE6WdE7kcr2-0nGeOezTpx_1dzu8FDCn3DJDjXg6C6DIkFkECPB2nsICQQ2wYFCCBSe9u6b7II_Cs54QmQWl_5yedaFQmMVRERURsunICi4sZHp-hXFDsg8-q4WPDN1ouBmW9qSkKb0ZmVqxTxWnpqV-IO2gEVH52KQAL3ccaWR_m1ZC3pZSjtnx0ozxFa4Cei1JupoGm8yK4imiGYBEnDFrU4zNcLiRBwOwAEYUsZk0ij5b-bL8wdZrnzFgMbP_ddRJafZmzhqzLH93EWJkhz9r93Cd8RzEOYNW8sMVsc-0hwPwqp1mhnYIrBcxMXkw71wcp0UVLCI154TKDC__HpI4b-EwtHh3hRsg77dd5_vNT4csT8GRFp3wD_azNe4sKRsBjnMw96vhm6A2ISE0Ib8pUACF5Jb5Fi70Dat63IS3BWZOeq1FzY9b64XaAZ6sTED3Uu5gOtN-x7tJMhM7xxaONdcHMRPg-2LkKsp1fVFRV_CdrZ2TBqoFPgKSuXy0

    Sequence Diagram for Deletion of User's Attributes

**Step 1:** The User requests the deletion of attributes invoking the Wallet Instance’s attribute deletion function.

> **Step 2:** The Wallet Instance collects all transaction data and shows the User the list of Relying Parties with which it has had interactions throughout the Wallet Instance lifecycle and are in possession of User's attributes. The Wallet Instance SHOULD filter the transaction logs so that only the Relying Parties which have had access to attributes uniquely identifying the User (e.g., the "Codice Fiscale") are shown.

Reviewer: ```suggestion
**Step 2:** The Wallet Instance collects all transaction data and shows the User the list of Relying Parties with which it has had interactions throughout the Wallet Instance lifecycle and are in possession of User's attributes. The Wallet Instance SHOULD filter the transaction logs so that only the Relying Parties which have had access to attributes uniquely identifying the User are shown.
```

**Step 3:** The User selects the target Relying Party for attributes deletion.

> **Steps 4 - 5:** The Wallet Instance obtains the Relying Party Entity Configuration at the Federation ./well-known/ endpoint. The URL or the Erasure Endpoint can be found inside the ``metadata.erasure_endpoint`` claim.

Reviewer: ```suggestion
**Steps 4 - 5:** The Wallet Instance obtains the Relying Party Entity Configuration at the Federation ./well-known/ endpoint. The URL or the Erasure Endpoint (``erasure_endpoint``) can be found inside the metadata parameter.
```

**Step 6:** The Wallet Instance logs the Erasure Request’s relevant information. These logs MUST include at least:
  * the date of request,
  * the Relying Party to which the request was made, 
  * the attributes requested to be removed.

> **Steps 7 - 8:** The Wallet Instance redirects the User to the Erasure Endpoint. It MUST also ensure that a callback mechanism to allow the User-Agent to notify the Wallet Instance (and thus the User) after the Erasure Response is present. This may be done e.g., by using an embedded User-Agent and inserting a ``redirect_uri`` in the form of a deep link scheme in the Erasure Request. Details on the Erasure Request can be found in :ref:`Erasure Request`.

Reviewer: ```suggestion
**Steps 7 - 8:** The Wallet Instance redirects the User to the Erasure Endpoint. It MUST also ensure that a callback mechanism to allow the User-Agent to notify the Wallet Instance (and thus the User) after the Erasure Response is present. Details on the Erasure Request can be found in :ref:`Erasure Request`.
```

.. note::
  
  The Relying Party web page will authenticate the User with an appropriate level of assurance using any method such as SPID/CIE or the PID presentation. The specific mechanism used for authentication is left to the Relying Party. Upon authenticating the User, the Relying Party MAY prompt the User to perform additional steps needed for the deletion of attributes, e.g., it might require the User to confirm the deletion operation.

**Step 9:** Upon successful authentication of the User the Relying Party MUST delete all attributes bound to the User in its possession. 

> **Step 10:** The Relying Party returns the Erasure Response in the form of an HTTP Response to the User-Agent and includes the ``redirect_uri`` if provided in the Erasure Request. Details on the Erasure Response can be found in :ref:`Erasure Response`.

Reviewer: ```suggestion
**Step 10:** The Relying Party returns the Erasure Response in the form of an HTTP Response to the User-Agent and includes the callback URL if provided in the Erasure Request. Details on the Erasure Response can be found in :ref:`Erasure Response`.
```

This Wallet Instance functionality allow the User to obtain from the Instance itself a list of all Relying Parties towards which attributes that can uniquely identify Users (e.g., the Codice Fiscale claim of the PID) have been presented. Subsequently the User is free to request deletion of all attributes presented to a Relying Party of its choice. Below the high level flow regarding this interaction is presented.   

.. figure:: ../../images/images/user's_data_deletion_flow.svg
    :figwidth: 100%
    :align: center
    :target: https://www.plantuml.com/plantuml/svg/RLBDZjD03BxdAQozS67t0Nf0ksABn0KX5iI5YvCuxOZfE8mzBNrxx2bfcBAtTFpv-_7NHr7CMWwnmwASog6dtE6WdE7kcr2-0nGeOezTpx_1dzu8FDCn3DJDjXg6C6DIkFkECPB2nsICQQ2wYFCCBSe9u6b7II_Cs54QmQWl_5yedaFQmMVRERURsunICi4sZHp-hXFDsg8-q4WPDN1ou7n5JePSfQ16Wljtxd1Zde_yam1LS-YB4emKgN9C9Gt-WJ6O7t2uR_ds1bxsV88OH86dndaWWHyf99XPX4ISYQVbU4-BpAsD5rCT57HFxHr0sUWo_IeaTRpwusbrBQi_JxjfIKnu-zwUAaa75GBtYtGToSn9oA-JM0duIDcdzXlWgwdUD0pSAuOaFg-kbeQMmyEK-U2JYZYGueYAvld_Y8RmSbotMsEujLUL8-_yml-ABidMZd3Z1qR_9h_7Az0cgdVnjaAlYTiQi9XWqZ2WadY6skX3HKwHJp2mJPDnmyc05L4BtgD0J_OYPHX8P2enjdJZGtk1QcDr_knzqrgrX--v65xvQkIwLSDxNAfvcsOBNak__Jj_OWtJ-KBrQ5xA8Vu1

>     Sequence Diagram for Deletion of User's Attribtes

Reviewer 2: Typo "Attribtes", change it to "Attributes"

**Step 1** The User requests the deletion of attributes invoking the Wallet Instance’s deletion function.

**Step 2** The Wallet Instance collects all transaction data and shows the User the list of Relying Parties with which it has had interactions throughout the Wallet Instance lifecycle and are in possession of User's attributes. The Wallet Instance SHOULD filter the transaction logs so that only the Relying Parties which have had access to attributes uniquely identifying the User (e.g., the "Codice Fiscale") are shown.

**Step 3** The User selects the target Relying Party for attributes erasure.

**Steps 4 - 5** The Wallet Instance obtains the Relying Party Entity Configuration at the Federation ./well-known/ endpoint. The URL or the Erasure Endpoint can be found inside the ``metadata.erasure_endpoint`` claim.

**Step 6** The RP logs the Erasure Request’s relevant information. These logs MUST include at least:
  * the date of request,

>   * the Relying Party to which the request was made, 

Reviewer 2: why the RP should also log its ID?

  * the attributes requested to be removed.

> **Steps 7 - 8** The Wallet Instance redirects the User to the Erasure Endpoint. It MUST also ensure that a call back mechanism to let the User-Agent notify the Wallet Instance (and thus the User) after the Erasure Response is present. This may be done e.g., by using an embedded User-Agent and inserting a ``redirect_uri`` in the form of a deep link scheme in the Erasure Request. Details on the Erasure Request can be found in :ref:`Erasure Request`.

Reviewer 2: It MUST also ensure a call back mechanism ...
Remove "that".

.. note::
  

>   The RP web page will authenticate the User with an appropriate level of assurance using any method such as SPID/CIE or the PID presentation. The specific mechanism used for authentication is left to the Relying Party. Upon having authenticated the User, the Relying Party MAY prompt the User to perform additional steps needed for the deletion of attributes, e.g., it might require the User to confirm the deletion operation.

Reviewer 2: Move "Upon having authenticated the User, the Relying Party MAY prompt the User to perform additional steps needed for the deletion of attributes, e.g., it might require the User to confirm the deletion operation." to Step 9.

**Step 9** Upon successful authentication of the User the Relying Party MUST delete all attributes bound to the User in its possession. 

> **Step 10** The RP returns the Erasure Response in the form of an HTTP Response to the User-Agent and includes the ``redirect_uri`` if provided in the Ereasur Request. Details on the Erasure Response can be found in :ref:`Erasure Response`.

Reviewer 2: Typo "Ereasur Request"

Wallet Attestation Issuance Request
.............................................

Further details on the Wallet Attestation Issuance Request are provided in the :ref:`Mobile Application Key Binding Request` section.

> The ``typ`` header of the Integrity Request JWT assumes the value ``war+jwt``.

Reviewer 3: ```suggestion
The ``typ`` header of the Integrity Request JWT assumes the value ``wp-war+jwt``.
```
